test/test_terraform_controller.py

```diff
@@ -0,0 +1,121 @@
+
+    @pytest.mark.parametrize('cloud', ['aws'])
+    def test_get_instances(self, mocker, tf_controller, cloud):
+        # Arrange
+        tf_controller.cloud = cloud
+
+        mock_popen = mocker.patch('os.popen', return_value=os._wrap_close)
+
+        mock_read = mocker.MagicMock(return_value='test_json')
+        os._wrap_close.read = mock_read
+
+        mock_loads = mocker.patch(
+            'json.loads',
+            return_value={'values': {'root_module': {'resources': 'test'}}},
+        )
+
+        mock_get_instances_cloud = mocker.MagicMock(return_value={'test_dict': 42})
+        if cloud == 'aws':
+            tf_controller.get_instances_aws = mock_get_instances_cloud
+
+        # Act
+        result = tf_controller.get_instances()
+
+        # Assert
+        mock_popen.assert_called_once_with('terraform show --json')
+        mock_read.assert_called_once()
+        mock_loads.called_once_with('test_json')
+        mock_get_instances_cloud.assert_called_once_with('test')
+        assert result == {'test_dict': 42}
+
+    def test_get_instances_aws(self, mocker, tf_controller):
+        # Arrange
+        resources = [
+            {
+                'address': 'a.aws_instance_test',
+                'values': {
+                    'id': 'test_id',
+                    'public_ip': 'test_ip',
+                    'public_dns': 'test_dns',
+                    'availability_zone': 'test_zone',
+                    'ami': 'test_ami',
+                },
+            },
+            {'address': 'a.not_an_instance'},
+        ]
+
+        instances_info_expected = {
+            'a.aws_instance_test': {
+                'instance_id': 'test_id',
+                'public_ip': 'test_ip',
+                'public_dns': 'test_dns',
+                'availability_zone': 'test_zone',
+                'ami': 'test_ami',
+                'username': 'test_user',
+            }
+        }
+
+        mock_get_username_by_instance_name = mocker.MagicMock(return_value='test_user')
+        self.tf_configurator.get_username_by_instance_name = mock_get_username_by_instance_name
+
+        # Act
+        result = tf_controller.get_instances_aws(resources)
+
+        # Assert
+        mock_get_username_by_instance_name.assert_called_once_with('aws_instance_test')
+        assert result == instances_info_expected
+
+    def test_destroy_resource(self, mocker, tf_controller):
+        # Arrange
+        mock_os_system = mocker.patch('os.system', return_value='')
+        test_resource_id = 'test_resource'
+        tf_destroy_resource = f'terraform destroy -target={test_resource_id}'
+
+        # Act
+        result = tf_controller.destroy_resource(test_resource_id)
+
+        # Assert
+        assert result is None
+        mock_os_system.assert_called_once_with(tf_destroy_resource)
+
+    def test_destroy_infra(self, mocker, tf_controller):
+        # Arrange
+        mock_os_system = mocker.patch('os.system', return_value='')
```

Comment:
`os.system` returns something if something failed while executing the command. If I don't set a `return_value`, the mock function will return something, so the code will exit because it thinks something failed.